File: Advanced Graph Problems/MST_Edge_Cost.py
import sys,math,cmath,random,os
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations,combinations
from io import BytesIO, IOBase
from decimal import Decimal,getcontext
 
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._file = file
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
 
# functions #
# MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(1,2**62)
def gcd(a,b):
    while b:
        a,b = b,a%b
    return a
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
II = lambda : int(sys.stdin.readline().strip())
LII = lambda : list(map(int, sys.stdin.readline().split()))
MI = lambda x : x(map(int, sys.stdin.readline().split()))
SI = lambda : sys.stdin.readline().strip()
SLI = lambda : list(map(lambda x:ord(x)-97,sys.stdin.readline().strip()))
LII_1 = lambda : list(map(lambda x:int(x)-1, sys.stdin.readline().split()))
LII_C = lambda x : list(map(x, sys.stdin.readline().split()))
MATI = lambda x : [list(map(int, sys.stdin.readline().split())) for _ in range(x)]
##
 
#String hashing: shclass, fenwick sortedlist: fsortl, Number: numtheory/numrare, SparseTable: SparseTable
#Bucket Sorted list: bsortl, Segment Tree(lp,selfop): SegmentTree, bootstrap: bootstrap, Trie: tries
#binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, BitArray: bitarray
#Combinatorics: pnc, Diophantine Equations: dpheq, DSU: DSU, Geometry: Geometry, FFT: fft, XOR_dict: xdict
#Persistent Segment Tree: perseg, Binary Trie: b_trie, HLD: hld, String funcs: sf, Segment Tree(lp): SegmentOther
#Graph1(dnc,bl): graphadv, Graph2(khn,sat): 2sat, Graph3(fltn,bprt): graphflatten, Graph4(ep,tp,fw,bmf): graphoth
#Graph5(djik,bfs,dfs): graph, Graph6(dfsin): dfsin, utils: utils, Persistent DSU: perdsu, Merge Sort Tree: sorttree
#2-D BIT: 2DBIT, MonoDeque: mono, nummat: matrix
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
# input_file = open(r'input.txt', 'r');sys.stdin = input_file
 
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
 
    def find(self, a):
        acopy = a
        while a != self.parent[a]:
            a = self.parent[a]
        while acopy != a:
            self.parent[acopy], acopy = a, self.parent[acopy]
        return a
 
    def union(self, a, b):
        self.parent[self.find(b)] = self.find(a)
 
 
class DisjointSetUnion:
    def __init__(self, n):
        self.parent = list(range(n))
        # self.s = set(self.parent)
        self.size = [1] * n
        self.flags = defaultdict(set)
        self.ans = {}
 
    def find(self, a):
        acopy = a
        while a != self.parent[a]:
            a = self.parent[a]
        while acopy != a:
            self.parent[acopy], acopy = a, self.parent[acopy]
        return a
 
    def union(self, a, b, w):
        a, b = self.find(a), self.find(b)
        if a != b:
            if self.size[a] < self.size[b]:
                a, b = b, a
            # self.s.remove(b)
            if len(self.flags[b])>len(self.flags[a]):
                self.flags[b],self.flags[a] = self.flags[a],self.flags[b]
            
            for j in self.flags[b]:
                if j in self.flags[a]:
                    self.flags[a].remove(j)
                    self.ans[j] = w
                else:
                    self.flags[a].add(j)
            self.flags[b].clear()
            self.parent[b] = a
            self.size[a] += self.size[b]
            return w
        return 0
 
    def set_size(self, a):
        return self.size[self.find(a)]
 
    def __len__(self):
        return len(self.s)
 
    def notfind(self, a):
        k = self.find(a)
        for j in self.s:
            if j!=k:
                return j
        return -1
    
def solve():
    n,m = LII()
    L = []
    ds = DisjointSetUnion(n)
    for _ in range(m):
        u,v,w = LII_1()
        L.append((w+1,u,v,_))
        ds.flags[u].add(_)
        ds.flags[v].add(_)
 
    L.sort()
    
    ans1 = 0
    for w,u,v,ind in L:
        ans1 += ds.union(u,v,w)
    
    ans = [0]*m
 
    for w,u,v,ind in L:
        # if ds.ans[ind]==w:
        ans[ind] = ans1+w-ds.ans[ind]
    
    print(*ans,sep='\n')
    #L1 = LII()
    #st = SI()
solve()

File: Advanced Graph Problems/MST_Edge_Cost.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace __gnu_pbds;
using namespace std;
 
#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define len(arr) arr.size()
#define f first
#define s second
#define pb push_back
#define all(x) x.begin(), x.end()
#define range(i, n) for (int i = 0; i < (n); ++i)
#define rangea(i, a, b) for (int i = (a); i < (b); ++i)
#define MOD1 998244353
#define MOD 1000000007
using vi = vector<int>;
using vii = vector<vector<int>>;
using pi = pair<int,int>;
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")
 
template <typename T>
using ordered_multiset = tree<
    std::pair<T, int>,
    null_type,
    std::less<std::pair<T, int>>,
    rb_tree_tag,
    tree_order_statistics_node_update
>;
template<typename T> istream& operator>>(istream& in, vector<T>& v) {for (auto& x : v) in >> x;return in;}
template<typename T1, typename T2> istream& operator>>(istream& in, pair<T1, T2>& p) {in >> p.first >> p.second;return in;}
template<typename T>
class is_printable {
    template<typename U> static auto test(int) -> decltype(cout << declval<U>(), true_type{});
    template<typename> static auto test(...) -> false_type;
public: static constexpr bool value = decltype(test<T>(0))::value;
};
template<typename T> enable_if_t<is_printable<T>::value> __print(const T& val) { cout << val; }
template<typename T1, typename T2> void __print(const pair<T1, T2>& p) {
    cout << "("; __print(p.first); cout << ":"; __print(p.second); cout << ")";
}
template<typename T> enable_if_t<!is_printable<T>::value> __print(const T& container) {
    cout << "{"; bool first = true;
    for (const auto& x : container) { if (!first) cout << ", "; __print(x); first = false; }
    cout << "}";
}
void print() { cout << "\n"; }
template<typename T, typename... Args>
void print(const T& t, const Args&... rest) {
    __print(t); if constexpr (sizeof...(rest)) cout << " "; print(rest...);
}
 
int gcd(int a, int b) {if (b == 0) return a;return gcd(b, a % b);}
int lcm(int a, int b) {return a / gcd(a, b) * b;}
int II(){int a;cin>>a;return a;}
string SI(){string s;cin>>s;return s;}
vi LII(int n){vi a(n);cin>>a;return a;}
 
// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree, Merge Sort Tree: sorttree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph, SortedList: sortl
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming
 
class DisjointSetUnion {
public:
    vector<int> parent, size;
    vector<set<int>> flags;
    map<int,int> ans;
 
    DisjointSetUnion(int n) : parent(n), size(n, 1) {
        iota(parent.begin(), parent.end(), 0);
        flags.resize(n);
    }
 
    int find(int a) {
        if (parent[a] != a) {
            parent[a] = find(parent[a]);
        }
        return parent[a];
    }
 
    int unite(int a, int b, int w) {
        // print(a,b);
        a = find(a);
        b = find(b);
        if (a != b) {
            if (size[a] < size[b]) swap(a, b);
 
            if(len(flags[a])<len(flags[b])){
                swap(flags[a],flags[b]);
            }
            for(auto j:flags[b]){
                if(flags[a].find(j)!=flags[a].end()){
                    flags[a].erase(j);
                    ans[j] = w;
                }
                else{
                    flags[a].insert(j);
                }
            }
            flags[b].clear();
 
            parent[b] = a;
            size[a] += size[b];
            return w;
        }
        return 0;
    }
 
    int set_size(int a) {
        return size[find(a)];
    }
 
    int notfind(int a) {
        int k = find(a);
        for (int i = 0; i < parent.size(); i++) {
            if (find(i) != k) {
                return i;
            }
        }
        return -1;
    }
};
 
void solve() {
    int n,m;
    cin>>n>>m;
 
    DisjointSetUnion ds(n);
 
    vector<tuple<int,int,int,int>> L;
    range(i,m){
        int u,v,w;
        cin>>u>>v>>w;
        L.emplace_back(w,u-1,v-1,i);
        ds.flags[u-1].insert(i);
        ds.flags[v-1].insert(i);
    }
 
    sort(all(L));
    
    int ans1 = 0;
    for(auto [w,u,v,ind]:L){
        ans1 += ds.unite(u,v,w);
    }

    vector<int> ans(m);
    // print(ds.ans);
    for(auto [w,u,v,i]:L){
        ans[i] = ans1-ds.ans[i]+w;
            // print("YES");
    }
 
    for(auto it:ans){
        print(it);
    }
 
}
 
int32_t main() {
    fastio
    int t=1;
    // cin >> t;
    while (t--) solve();
    return 0;
}